kernel/siw_cm.c

```diff
@@ -1264,6 +1418,48 @@ static void siw_cm_llp_state_change(struct sock *sk)
 	orig_state_change(sk);
 }
 
+void siw_rtr_complete(struct siw_qp *qp)
+{
+	struct siw_cep	*cep;
+
+	cep = qp->cep;
+
+	siw_enter_rts(cep, qp);
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 15, 0)
+void siw_rtr_data_ready(struct sock *sk, int flags)
+#else
+void siw_rtr_data_ready(struct sock *sk)
+#endif
+{
+	struct siw_cep		*cep;
+	struct siw_qp		*qp = NULL;
+	read_descriptor_t	rd_desc;
+
+	read_lock(&sk->sk_callback_lock);
+
+	cep = sk_to_cep(sk);
+	if (!cep) {
+		WARN_ON(1);
+		goto out;
+	}
+	qp = sk_to_qp(sk);
+
+	memset(&rd_desc, 0, sizeof(rd_desc));
+	rd_desc.arg.data = qp;
+	rd_desc.count = 1;
+	if (down_read_trylock(&qp->state_lock)) {
```

Comment:
I don't think that we actually need the state lock. We have already transitioned to RTS. This will just signal this with an upcall indicating the RTR.